flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala

```diff
@@ -0,0 +1,331 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.flink.table.runtime.aggregate
+
+import java.lang.{Long => JLong}
+import java.util.{ArrayList, List => JList}
+
+import org.apache.flink.api.common.state._
+import org.apache.flink.api.java.typeutils.RowTypeInfo
+import org.apache.flink.configuration.Configuration
+import org.apache.flink.runtime.state.{FunctionInitializationContext, FunctionSnapshotContext}
+import org.apache.flink.streaming.api.checkpoint.CheckpointedFunction
+import org.apache.flink.streaming.api.functions.ProcessFunction
+import org.apache.flink.table.functions.{Accumulator, AggregateFunction}
+import org.apache.flink.types.Row
+import org.apache.flink.util.{Collector, Preconditions}
+import org.apache.flink.table.TableAPIConfigConstant._
+
+import scala.collection.mutable
+import scala.collection.mutable.PriorityQueue
+
+/**
+  * Process Function for ROWS clause bounded OVER window
+  *
+  * @param aggregates           the list of all [[org.apache.flink.table.functions.AggregateFunction]]
+  *                             used for this aggregation
+  * @param aggFields            the position (in the input Row) of the input value for each aggregate
+  * @param forwardedFieldCount  the count of forwarded fields.
+  * @param aggregationStateType the row type info of aggregation
+  * @param inputType            the row type info for input data
+  * @param processingOffset     the offset of processing
+  * @param isRowTimeType        It is a tag that indicates whether the time type is rowTimeType
+  */
+class RowsClauseBoundedOverProcessFunction(
+    private val aggregates: Array[AggregateFunction[_]],
+    private val aggFields: Array[Int],
+    private val forwardedFieldCount: Int,
+    private val aggregationStateType: RowTypeInfo,
+    private val inputType: RowTypeInfo,
+    private val processingOffset: Int,
+    private val isRowTimeType: Boolean)
+    extends ProcessFunction[Row, Row] with CheckpointedFunction {
+
+  Preconditions.checkNotNull(aggregates)
+  Preconditions.checkNotNull(aggFields)
+  Preconditions.checkArgument(aggregates.length == aggFields.length)
+
+  private var output: Row = _
+  private var accumulators: Row = _
+  private var sameTimeDataBuffer: JList[Row] = _
+
+  // save last expired data
+  private var lastExpiredRow: Row = _
+  // persistent last expired data for recovery
+  private var lastExpiredRowState: ListState[Row] = null
+
+  // for event-time, sorted the data by timestamp
+  private var bufferedRowPriorityQueue: PriorityQueue[(Long, Long)] = _
+  // for proc-time
+  private var bufferedRowList: JList[Long] = _
+  // persistent buffered data for recovery
+  private var bufferOperatorState: ListState[Long] = null
+
+
+  // save incremental calculation accumulator
```

Comment:
A backend state which is used to materialize the time stamp. For event-time processing, it saves the bufferedRowPriorityQueue, while for proc-time, it saves bufferedRowList.